### PR TITLE
Log that steps are cached while running pipeline

### PR DESCRIPTION
**Pull request by alex-zenml**
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes

Introspected the metadata store while pipeline steps are being executed by tfx.

### CODE REVIEW

File: src/zenml/orchestrators/utils.py
Comment (htahir1) [inline src/zenml/orchestrators/utils.py:69]: Instead of `zenml-pipeline_parameter_name` I would use the constants defined in `BaseMetadataStore`

```
from zenml.metadata_stores.base_metadata_store import INTERNAL_EXECUTION_PARAMETER_PREFIX , PARAM_PIPELINE_PARAMETER_NAME

step_name_property = INTERNAL_EXECUTION_PARAMETER_PREFIX + PARAM_PIPELINE_PARAMETER_NAME,
```

File: src/zenml/orchestrators/utils.py
Comment (htahir1) [inline src/zenml/orchestrators/utils.py:70]: Why do we need to do `1:-1` here? If we want to strip it we can use `.strip()`? Also we should add exception handling here if this exec_property does not exist (Maybe raise a KeyError)

File: src/zenml/orchestrators/utils.py
Comment (htahir1) [inline src/zenml/orchestrators/utils.py:76]: Any way we can avoid these `type: ignore` here? @schustmi perhaps any tips?

File: src/zenml/orchestrators/utils.py
Comment (htahir1) [inline src/zenml/orchestrators/utils.py:55]: I would say `Returns the caching state of a step`

File: src/zenml/orchestrators/utils.py
Comment (htahir1) [inline src/zenml/orchestrators/utils.py:100]: Can we also add some more details? Like the actual function name instead? So e.g. if you pass in `importer_mnist` and the pipeline parameter name is `importer` then we can say: `f"Using cached version for {pipeline_step_name} [{step_name}] from run_id: {run_id}"`

File: src/zenml/orchestrators/utils.py
Comment (htahir1) [inline src/zenml/orchestrators/utils.py:94]: Rename this to: `pipeline_step_name`? This is the name defined in the @pipeline

File: src/zenml/orchestrators/utils.py
Comment (htahir1) [inline src/zenml/orchestrators/utils.py:64]: What if the stack changes in the middle? i.e. a pipeline is long-running (.e.g two hours). While its running user changes stack.

We should probably not get the active stack here but the stack used by the pipeline. This will be in exec_properties I believe when @bcdurak branch is merged

**Review by htahir1 — CHANGES_REQUESTED** [review]
Ok this is really good -> It will be super useful in the logs.

Apart from my comments, I do have a slight concern that this causes one extra connection (request) to be made to the MLMD that will slow things down a bit for every run. This is not a concern right now but i wonder how it impacts performance overall when things scale.. I do not have a better suggestion right now to fix it other than to dive deeper into the `launch()` code that produces the `execution_info` and sort of put the information there (it already exists in the `launch` code but somehow TFX never puts it in `execution_info`...

File: src/zenml/orchestrators/utils.py
Comment (alex-zenml) [reply src/zenml/orchestrators/utils.py:70]: You can't actually use `.strip()` on it. Prior to doing the slicing, we have `'"step_one"'`. For some reason `tfx` packages it that way. I'd be happy if there was a Python method that removed the extra quotation marks, but I don't know it. Any suggestions @htahir1?

File: src/zenml/orchestrators/utils.py
Comment (alex-zenml) [reply src/zenml/orchestrators/utils.py:76]: Yeah actually we could define a mypy ignore file where we could define separately which lines or statements could be ignored for certain classes of errors, but I think that's a) a slippy slope and b) too many things change too often in our codebase for that to work.

The problem (I think) is how `tfx` type annotate their outputs, which then we have to handle. One way of getting around it would be (at the top of the file) to turn off the strict errors for whenever we're not handling a potential `None` value that comes out of `tfx` in that chain. Not sure whether that's a good idea or not.

File: src/zenml/orchestrators/utils.py
Comment (htahir1) [reply src/zenml/orchestrators/utils.py:70]: I think its a json string so you can do `import json; json.loads(execution_info.exec_properties[step_property_name])`

Comment (schustmi) [reply src/zenml/orchestrators/utils.py:70]: You can optionally pass in characters to strip, so `.strip(")` would probably work in this case.
I do however agree that we should use `json.loads` instead, as we (and not actually tfx) put them in the metadata store using `json.dumps` in the first place.

File: src/zenml/orchestrators/utils.py
Comment (schustmi) [reply src/zenml/orchestrators/utils.py:76]: I think the ignores here come from the fact that
- `metadata_store.get_pipeline(...)` returns an optional pipeline, so we would have to extract that and then check if the pipeline actually exists
- `pipeline_run_name` probably isn't a string, so we'd have to check beforehand to see that it's actually a string

So it would make the code a few lines longer due to these extra checks but at the same time safer and we can raise more explicit errors

**Review by schustmi — COMMENTED** [review]
I agree with all of @htahir1's feedback, and would also suggest looking into the execution info further to see if we can find a way to check if pipelines are cached. After a quick glance there seems to be no explicit attribute specifying it, but there is lots of stuff like artifact uris that we could potentially "abuse" to figure out if the step was cached. Also happy to pair on that if you'd like

File: src/zenml/orchestrators/utils.py
Comment (alex-zenml) [reply src/zenml/orchestrators/utils.py:70]: Done

File: src/zenml/orchestrators/utils.py
Comment (alex-zenml) [reply src/zenml/orchestrators/utils.py:76]: Done

**Comment by alex-zenml** [comment]
@htahir1 @schustmi these latest commits merely address fixes for the current solution. I'll take a look at doing things with only the `execution_info` now as an alternative.

File: src/zenml/orchestrators/utils.py
Comment (htahir1) [inline src/zenml/orchestrators/utils.py:127]: Maybe I am misunderstanding the flow here, but wouldn't `step_name` not exist if it goes into the `else` condition above? Perhaps it might make sense to just but this log in the `if` , and with the `else` raise an exception?

File: src/zenml/orchestrators/utils.py
Comment (htahir1) [inline src/zenml/orchestrators/utils.py:59]: Might be worth writing a unit few tests for this function to make sure it gets out the right thing

**Review by htahir1 — CHANGES_REQUESTED** [review]
Alright thanks for implementing the feedback. I left a few more notes, and looking forward to the alternate implementation when you have it (We can do that in another PR after merging this?)

File: src/zenml/orchestrators/utils.py
Comment (alex-zenml) [reply src/zenml/orchestrators/utils.py:59]: I probably would have to mock up lots of pieces to make this work fast: the Repository, the metadata_store, specific pipelines and steps. Do you see any way to do this unit test without doing all of that? (Also maybe @schustmi you did some of this mocking already of certain components given that you recently wrote a bunch of tests across stack components...)

Comment (schustmi) [reply src/zenml/orchestrators/utils.py:59]: Mhm yes certainly not an easy test, but we can maybe get away with not doing any mocking at all:
- Run the same pipeline (with cache enabled) twice inside a `clean_repo`
- Create an `execution_info` object with the right pipeline/run/step names
- Check if it returns `False` for the first run and `True` for the second one

File: .gitignore
Comment (schustmi) [inline .gitignore:184]: Maybe we can remove this line entirely now that sphinx is no longer used?

File: src/zenml/orchestrators/utils.py
Comment (alex-zenml) [reply src/zenml/orchestrators/utils.py:59]: Ok so it'd be an integration test, then, and not a unit test. Yep that's easy to do.

Comment (alex-zenml) [reply src/zenml/orchestrators/utils.py:59]: Actually I misspoke. Not an integration test, but somewhere in between. I saw how you've been using `clean_repo` in other tests and modelled something around that in my implementation.

File: src/zenml/orchestrators/utils.py
Comment (alex-zenml) [reply src/zenml/orchestrators/utils.py:82]: Waiting on the branch to be merged to work on this.

Comment (alex-zenml) [reply src/zenml/orchestrators/utils.py:82]: Managed to implement this, though it's a bit messy due to how it's stored in the `ExecutionInfo` object. Just need to get the test amended to this new pattern which is proving a bit fiddly, but it seems doable.

**Comment by htahir1** [comment]
@alex-zenml  Let's re-open this branch as we reverted before the release

**Pull request merged**
